src/witness/memory.rs

```diff
@@ -81,13 +95,19 @@ impl SafeMemory {
 
     /// Writes a u32 to the specified memory offset
     pub fn write_u32(&mut self, ptr: usize, num: u32) {
-        let buf = unsafe { self.memory.data_unchecked_mut() };
+        let store = self.store.write().unwrap();
+        let view = self.memory.view(&*store);
+
+        let buf = unsafe { view.data_unchecked_mut() };
         buf[ptr..ptr + std::mem::size_of::<u32>()].copy_from_slice(&num.to_le_bytes());
     }
 
     /// Reads a u32 from the specified memory offset
     pub fn read_u32(&self, ptr: usize) -> u32 {
-        let buf = unsafe { self.memory.data_unchecked() };
+        let store = self.store.read().unwrap();
```

Comment:
I suspect there should be a clean way to refactor these repeating lines of read and write locking a memory view into a helper function

Reply:
still wanna take a stab at this or do we leave as-is?